src/link.c

```diff
@@ -1124,9 +1124,10 @@ static void link_transfer_cancel_handler(ASYNC_OPERATION_HANDLE link_transfer_op
     {
         pending_delivery->on_delivery_settled(pending_delivery->callback_context, pending_delivery->delivery_id, LINK_DELIVERY_SETTLE_REASON_CANCELLED, NULL);
     }
-    free(pending_delivery);
     
     (void)singlylinkedlist_remove_if(((LINK_HANDLE)pending_delivery->link)->pending_deliveries, remove_pending_delivery_condition_function, pending_delivery);
+
+    async_operation_destroy(link_transfer_operation);
```

Comment:
@dcristoloveanu , based on our conversation I understand the async_operation_destroy shall be invoked from within the cancellation handler.
Is that understanding correct?

I see uamqp doesn't call async_operation_cancel anywhere yet. We need to make sure this is well verified when it starts to use it.

Reply:
Your understanding is correct, thanks for fixing this.

---
In reply to: [153641131](https://github.com/Azure/azure-uamqp-c/pull/202#discussion_r153641131) [](ancestors = 153641131)